index.d.ts

```diff
@@ -0,0 +1,317 @@
+		 */
+
+		test(regex: RegExp): boolean
+	}
+
+	namespace Parser {
+		interface ParsedResult {
+			browser: Details;
+			os: OSDetails;
+			platform: PlatformDetails;
+			engine: Details;
+		}
+
+		interface Details {
+			name?: string;
+			version?: Array<{index: number, input: string} | boolean | string | any>;
+		}
+
+		interface OSDetails extends Details {
+			versionName?: string;
+		}
+
+		interface PlatformDetails {
+			type?: string;
+			vendor?: string;
+			model?: string;
+		}
+
+		type BrowserDetails = Details;
+		type EngineDetails = Details;
+
+		interface checkTree {
+			[key: string]: any;
+		}
+	}
+
+	class Utils {
+		/**
+		 * Get first matched item for a string
+		 * @param {RegExp} regexp
+		 * @param {String} ua
+		 * @return {Array|{index: number, input: string}|*|boolean|string}
+		 */
+		static getFirstMatch(regexp: RegExp, ua: string): Array<{index: number, input: string} | boolean | string | any>;
+		/**
+		 * Get second matched item for a string
+		 * @param regexp
+		 * @param {String} ua
+		 * @return {Array|{index: number, input: string}|*|boolean|string}
+		 */
+		static getSecondMatch(regexp: RegExp, ua: string): Array<{index: number, input: string} | boolean | string | any>;
+		
+		/**
+		 * Match a regexp and return a constant or undefined
+		 * @param {RegExp} regexp
+		 * @param {String} ua
+		 * @param {*} _const Any const that will be returned if regexp matches the string
+		 * @return {*}
+		 */
+		static matchAndReturnConst(regexp: RegExp, ua: string, _const: any): any | undefined;
+		
+		/**
+		 * Retrieves Windows commercial name from NT Core version name
+		 * @param {string} version
+		 * @returns {string | undefined}
+		 */
+		static getWindowsVersionName(version: string): string | undefined;
+		
+		/**
+		 * Get version precisions count
+		 *
+		 * @example
+		 *	 getVersionPrecision("1.10.3") // 3
+		 *
+		 * @param {string} version
+		 * @return {number}
+		 */
+		static getVersionPrecision(version: string): number
+		
+		/**
+		 * Calculate browser version weight
+		 *
+		 * @example
+		 *	 compareVersions('1.10.2.1',	'1.8.2.1.90')	// 1
+		 *	 compareVersions('1.010.2.1', '1.09.2.1.90');	// 1
+		 *	 compareVersions('1.10.2.1',	'1.10.2.1');	// 0
+		 *	 compareVersions('1.10.2.1',	'1.0800.2');	// -1
+		 *	 compareVersions('1.10.2.1',	'1.10',	true);	// 0
+		 *
+		 * @param {String} versionA versions versions to compare
+		 * @param {String} versionB versions versions to compare
+		 * @param {boolean} [isLoose] enable loose comparison
+		 * @return {Number} comparison result: -1 when versionA is lower,
+		 * 1 when versionA is bigger, 0 when both equal
+		 */
+		static compareVersions(versionA: string, versionB: string, isLoose?: boolean): number;
+		
+		/**
+		 * Array::map polyfill
+		 *
+		 * @param	{Array} arr
+		 * @param	{Function} iterator
+		 * @return {Array}
+		 */
+		static map(arr: Array<any>, iterator: Function): Array<any>
+	}
+
+	export = Bowser;
```

Comment:
@alexandercerutti, what if we put it like that?

```
export = Bowser;
export as namespace Bowser;
```

At least, it seems working for React: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L45

Reply:
Tomorrow I'll test on the real project, but following this post, it seems it may work: https://stackoverflow.com/questions/44847749/explain-export-and-export-as-namespace-syntax-in-typescript

Reply:
Thank you 🙏

Reply:
I tried to convert the whole thing to React-approach like:

```javascript
export = Bowser;
export as namespace Bowser;

declare namespace Bowser {
    export class Bowser { ... }
}
```

as to use what you said, a namespace is needed to be declared.

By doing this and `import Bowser from "bowser"`, we get a module that is required to be used as "Bowser.Bowser".
So, we can do `import { Bowser } from "bowser"` or look for another solution.

Reply:
Okay so, I found two other ways. The key here is to no use `declare module` but just `declare class Bowser`. I also extracted the namespace and the parser class and declared a namespace called NSParser, which includes Parser class.

Then we can do like suggested by you previously or add a leading `export` to declare for both namespace and bowser class.

like this:

```javascript
declare class Bowser { ... }
declare namespace ParserNS {
    interface ParsedResult { ... }
    // ...
    class Parser { ... }
}
```

If we do as you suggested, we export only Bowser.
If we set leading `export`, we can also export NSParser as types (the question is: **might they be needed by who uses the lib?**). Moreover, that might be a problem with this option, one might access directly to the Parser class. I don't think this is possible.

Reply:
Mhhh, I think I wasted half working day to test a good way, a no one, even yours, work.
The best ones work in VSCode but once the code is converted from TS to JS, it fails.

Reply:
I found [this StackOverflow answer](https://stackoverflow.com/a/35706271) written by Ryan Cavanaugh, one or maybe the main maintainer of Typescript, that reports this:

> If you're trying to import a function or class, you should use
```javascript
import express = require('express');
```

And if you think, the exported version of Bowser, is a function, not a class or a module.
Or better, this header / wrapper exports bowser as UMD, a mix of AMD and CommonJS.

```javascript
!function(e, t) {
    "object" == typeof exports && "object" == typeof module ? module.exports = t() : "function" == typeof define && define.amd ? define([], t) : "object" == typeof exports ? exports.bowser = t() : e.bowser = t()
}(this, function() {
    ...<bowser gets defined in here>...
});
```

It seems like the first provided way, with `import Bowser = require("bowser")`, is the best way.
Let me ask you a thing: is this library meant to be used also in node.js?

Reply:
So, I think we'll have to opt for this way and use that.

Reply:
Ok, if you think so, let's do this. We always can make changes later if any needed. I'm lost again: is the current `.d.ts` file stable one then or any changes you want to make? :)

Reply:
I think and I've tried 😂 this should be the stable version, but the only thing I don't know yet is if through these typings, users could maybe access to Parser as definition even if they cannot in real. I have to test this aspect.

I think that in maximum 5 days, the v2.0.0 stable will be ready.
I'm going also to edit the README.
Have you seen the other changes?

Reply:
I've updated the Readme. But what is "Impression"?

Reply:
If people are using typescript through babel `import Bowser = require('bowser');` is not supported, https://babeljs.io/docs/en/babel-plugin-transform-typescript

Reply:
Mmh, in the project i'm working on at work, I'm using Typescript + Webpack + Babel but not sure the one you linked. In that project the only way to make it work is through that type of import @JaceHensley

Reply:
If you are using babel-loader then it does use that babel-plugin, probably through [this](https://babeljs.io/docs/en/babel-preset-typescript#docsNav)

I don't think the wrapping `declare module "bowser" {` is needed, I think that's only needed for 3rd party typings. Also I think you have to do `declare class` instead of just `class`, from [here](https://www.typescriptlang.org/docs/handbook/declaration-files/by-example.html#classes).

So maybe something like

```
declare class Bowser {
  ...
}

declare class Utils {
  ...
}

export {Bowser, Utils, ...}
```

This actually seems like a pretty good guide in general https://www.typescriptlang.org/docs/handbook/declaration-files/by-example.html

Reply:
I agree. Moreover, Utils class is useless: I integrated into the typings mistakenly, so we don't need to export it and neither include it into typings.

I'm using babel-loader and I have no problems with that version.
Anyway, the problem is always the same: while trying different ways, I can get VSCode to recognize types but when using Bowser and compiling them from ts to js, I get, for example, "Cannot read getParser of undefined" or such.

Anyway, I found a way that should work @lancedikson: I followed again what you said about React typings and thought: "since Bowser exports only static methods, we may think it as a namespace with functions, just as React does".

So I came up with this (shortened version)

```javascript
export = Bowser;
export as namespace Bowser;

declare namespace Bowser {
    /**
     * Creates a Parser instance
     * @param {string} UA - User agent string
     * @param {boolean} skipParsing
     */

    function getParser(UA: string, skipParsing?: boolean): Parser.Parser

    /**
     * Creates a Parser instance and runs Parser.getResult immediately
     * @param UA - User agent string
     * @returns {Parser.ParsedResult}
     */

    function parse(UA: string): Parser.ParsedResult
}

declare namespace Parser {
    class Parser {
        constructor(UA: string, skipParsing?: boolean);
        getBrowser(): Parser.BowserDetails;
    }

    interface BrowserDetails {
        name?: string,
        version?: string,
    }
}
```

To import it, we have to write in our app `import * as Bowser from "bowser";` and then we can access to its methods easily, like `Bowser.getParser()`.

> Please note this is an incomplete version written here just to let you try it.
> I'll commit the complete version if this version will work also for you.

@lancedikson @JaceHensley can you try?

Reply:
This worked for me, it's odd that it can't be a class and can't be exported via `export default Bowser` but I'm not really sure why that is.

Reply:
types don't make that difference if it is a class or not. The right thing to watch is if
- VSCode see the typings and
- the exported result is compliant with the real definition.

As I said, with many types of typings structure, I got the first but not the second.

Reply:
Yeah I agree that creating types is harder than it should be, with the snippet above it does seem to work in VSCode and when I build the app with webpack (ts-loader)

Reply:
Okay, so I'm waiting for the approval of @lancedikson and then I'm going to do commit them.

Reply:
Yeah, let's commit that, I guess. I haven't tried it, but if you guys checked that and it worked, so let's have it merged. We can always release a hotfix later if something's missing.

Btw, I've found this [dts-gen](https://github.com/Microsoft/dts-gen) tool, but it doesn't seem working for me if I use it for bowser. I tried to get rid of `import/export` statements in order to make it be able to run through the files without babel, but it gave me just a file with declaration of `Bowser` class and a couple of its methods, no more.

Reply:
Perfect! Tomorrow I'm going to commit that since the full source is on my pc at work.

I never used that tool but by just reading its readme, it seems not to be that reliable to me.

Reply:
Okay, I'm going to push the changes.
@lancedikson, please notice that the comments for `Parser.satisfies` refer to a Bowser method that does not exist anymore and it still use Bowser contructor. I took all methods comments from the code.

```javascript
/**
 * Check if parsed browser matches certain conditions
 *
 * @param {checkTree} checkTree It's one or two layered object,
 * which can include a platform or an OS on the first layer
 * and should have browsers specs on the bottom-laying layer
 *
 * @returns {Boolean|undefined} Whether the browser satisfies the set conditions or not.
 * Returns `undefined` when the browser is no described in the checkTree object.
 *
 * @example
 * const browser = new Bowser(UA);
 * if (browser.check({chrome: '>118.01.1322' }))
 * // or with os
 * if (browser.check({windows: { chrome: '>118.01.1322' } }))
 * // or with platforms
 * if (browser.check({desktop: { chrome: '>118.01.1322' } }))
 */
```